swift/ql/test/library-tests/dataflow/taint/url.swift

```diff
@@ -41,19 +59,36 @@ func taintThroughURL() {
 	let urlTainted = URL(string: tainted)!
 
 	sink(arg: urlClean)
-	sink(arg: urlTainted) // $ tainted=39
+	sink(arg: urlTainted) // $ tainted=57
+	sink(arg: urlTainted.absoluteURL) // $ tainted=57
+	sink(arg: urlTainted.baseURL) // $ tainted=57
+	sink(string: urlTainted.fragment!) // $ tainted=57
+	sink(string: urlTainted.host!) // $ tainted=57
+	sink(string: urlTainted.lastPathComponent) // $ tainted=57
+	sink(string: urlTainted.path) // $ tainted=57
+	sink(string: urlTainted.pathComponents[0]) // $ tainted=57
+	sink(string: urlTainted.pathExtension) // $ tainted=57
+	sink(int: urlTainted.port!) // $ tainted=57
+	sink(string: urlTainted.query!) // $ tainted=57
+	sink(string: urlTainted.relativePath) // $ tainted=57
+	sink(string: urlTainted.relativeString) // $ tainted=57
+	sink(string: urlTainted.scheme!) // $ tainted=57
+	sink(string: urlTainted.standardized) // $ tainted=57
+	sink(arg: urlTainted.standardizedFileURL) // $ tainted=57
+	sink(string: urlTainted.user!) // $ tainted=57
+	sink(string: urlTainted.password!) // $ tainted=57
 
 	sink(arg: URL(string: clean, relativeTo: nil)!)
-	sink(arg: URL(string: tainted, relativeTo: nil)!) // $ tainted=39
+	sink(arg: URL(string: tainted, relativeTo: nil)!) // $ tainted=57
 	sink(arg: URL(string: clean, relativeTo: urlClean)!)
-	sink(arg: URL(string: clean, relativeTo: urlTainted)!) // $ tainted=39
+	sink(arg: URL(string: clean, relativeTo: urlTainted)!) // $ tainted=57
```

Comment:
I think we could do with at least one test for a field after `init(string:relativeTo:)`.

Reply:
This actually made me think more detailedly about this constructor, and after testing it, the conclusion is:

* if the `string:` argument is an absolute URL, basically it acts as the 1-argument constructor and `relativeTo:` is ignored (but `baseURL` is tainted by it).
* if the `string:` argument is a relative path:
  * the path, query, and fragment related fields are tainted by it.
  * all the other fields are tainted by `relativeTo:`.

I adjusted the CSV rows to model this behavior as closely as possible, but of course the conditional logic is not handled. If we wanted more precise models we would need to switch to regular CodeQL instead, and apply some heuristics to the first argument to determine whether it's an absolute URL or a relative path.

What's your opinion about that?